### DISCUSSION TITLE
Can't get .pdf exports working with VS Code & Chromium snaps/flatpaks

### DISCUSSION THREAD

**Discussion by frankhuurman** [marp-team/marp#491] [Q&A] [answered]
I saw there were a couple of comparable issues with snap packages and not being able to find or launch Chromium to export to .pdf but I couldn't find an answer to it.

I've got both VS Code and Chromium installed as snaps as seen below:
![image](https://github.com/marp-team/marp/assets/19980966/193996ad-c0ef-4758-b880-6f8fc28f6169)

but if I try to export this test slide:

```md
---
marp: true
---

# test
```

I get the well known:
```
Failure to export: [CLIError] Failed converting Markdown. (Failed to launch the browser process!
TROUBLESHOOTING: https://pptr.dev/troubleshooting
).
```

I've tried to point marp to the executable like this:
```md
---
marp: true
markdown.marp.chromePath: "/snap/chromium/current/usr/lib/chromium-browser/chrome"
---

# test
```

But even though I can launch that binary in the terminal, Marp still throws the same error when exporting.

Any other solution I can try to get it working?

**Comment by frankhuurman** [comment]
I've removed the chromium snap and installed the chromium flatpak instead and after restarting VS Code I now get a different error:

![image](https://github.com/marp-team/marp/assets/19980966/88774dbe-95d6-4d77-9bbe-f26470b18868)

I tried to use the chromium flatpak binary path as marp chromePath but that also doesn't work.

edit: Also installed the Edge flatpak but doesn't work either. Maybe I just need a non-snap version of VS Code?
edit 2: Nope, flatpak version of VS Code also isn't able to pick up on the browsers.

Unsure what would fix it currently.

**Reply by yhatt** [reply]
Marp cannot find out Flatpak binary automatically because [Google's `chrome-launcher` module that Marp is depending does not consider about Flatpak binaries](https://github.com/GoogleChrome/chrome-launcher/blob/main/src/chrome-finder.ts). Also in that case, you should specify the flatpak executable path [by the VS Code preference](https://github.com/orgs/marp-team/discussions/491#discussioncomment-7927971).

**Comment by yhatt** [comment, marked as answer]
> I've tried to point marp to the executable like this:
> 
> ```
> ---
> marp: true
> markdown.marp.chromePath: "/snap/chromium/current/usr/lib/chromium-browser/chrome"
> ---
> 
> # test
> ```

@frankhuurman `markdown.marp.chromePath` is [the setting ID for VS Code](https://code.visualstudio.com/docs/getstarted/settings), not part of Marp's Markdown directives.

You can find the correct place to set the Chromium executable path in the VS Code preferences.

![Settings - Visual Studio Code 2023_12_22 22_42_01](https://github.com/marp-team/marp/assets/3993388/28d3dc42-4578-442b-853c-6a0b218cea22)

**Reply by frankhuurman** [reply]
Thanks! Got it working :smile: 
Not sure where I got the example to use chromePath in the markdown directives but good to know where to find the correct one.

For anyone coming across this ticket in the future and has issues with exporting on Ubuntu 22.04 (or similar):
Using the Snap versions of VS Code & Chromium and setting `markdown.marp.chromePath` in VS Code settings to `/snap/chromium/current/usr/lib/chromium-browser/chrome` made it possible to export to .pdf for me.

**Reply by yhatt** [reply]
Before trying that, please be aware that directly using the binary specifically designed for the snap package means completely bypassing [snap's sandbox security model](https://ubuntu.com/core/services/guide/snaps-intro), which may pose potential risks.

Or you might want to consider switching either VS Code, Chrome/Chromium, or both to their non-snap versions.

Related:

- https://github.com/marp-team/marp-cli/issues/287

**Reply by frankhuurman** [reply]
I was aware of the security risks but I appreciate the warning, ゆき :smile:

**Reply by HappiePlant** [reply]
Has anyone figured out the flatpak path?
